mlem/contrib/venv.py

```diff
@@ -0,0 +1,187 @@
+        if env_type == "pip":
+            env_exe = os.path.join(env_dir, "Scripts", "python.exe")
+        elif env_type == "conda":
+            env_exe = os.path.join(env_dir, "python.exe")
+    else:
+        env_exe = os.path.join(env_dir, "bin", "python")
+    return env_exe
+
+
+def run_in_subprocess(cmd: List[str], error_msg: str, check_output=False):
+    try:
+        if check_output:
+            return subprocess.check_output(cmd)
+        return subprocess.run(cmd, check=True)
+    except (
+        FileNotFoundError,
+        subprocess.CalledProcessError,
+        subprocess.TimeoutExpired,
+    ) as e:
+        raise MlemError(f"{error_msg}\n{e}") from e
+
+
+class EnvBuilder(MlemBuilder):
+    """MlemBuilder implementation for building virtual environments"""
+
+    type: ClassVar = "env"
+
+    target: Optional[str] = "venv"
+    """Name of the virtual environment"""
+
+    @abstractmethod
+    def create_virtual_env(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def get_installed_packages(self, env_dir: str):
+        raise NotImplementedError
+
+
+class VenvBuilder(EnvBuilder):
+    class Config:
+        arbitrary_types_allowed = True
+        exclude = {"context"}
+
+    type: ClassVar = "venv"
+
+    no_cache: bool = False
+    """Disable cache"""
+    current_env: bool = False
+    """Whether to install in the current virtual env, must be active"""
+    context: Optional[SimpleNamespace] = None
+    """context for the virtual env"""
+
+    def create_virtual_env(self):
+        env_spec = venv.EnvBuilder(with_pip=True)
+        env_dir = os.path.abspath(self.target)
+
+        self.context = env_spec.ensure_directories(env_dir)
+        env_spec.create_configuration(self.context)
+        env_spec.setup_python(self.context)
+        if env_spec.with_pip:
+            env_spec._setup_pip(  # pylint: disable=protected-access
+                self.context
+            )
+        if not env_spec.upgrade:
+            env_spec.setup_scripts(self.context)
+            env_spec.post_setup(self.context)
+
+    def get_installed_packages(self, env_dir):
+        env_exe = get_python_exe_in_virtual_env(env_dir)
+        return run_in_subprocess(
+            [env_exe, "-m", "pip", "freeze"],
+            error_msg="Error running pip",
+            check_output=True,
+        )
+
+    def build(self, obj: MlemModel):
+        if self.current_env:
+            if (
+                os.getenv("VIRTUAL_ENV") is None
+                or sys.prefix == sys.base_prefix
+            ):
+                raise MlemError("No virtual environment detected.")
+            echo(EMOJI_PACK + f"Detected the virtual env {sys.prefix}")
+            env_dir = sys.prefix
+            env_exe = get_python_exe_in_virtual_env(env_dir)
+        else:
+            echo(EMOJI_PACK + f"Creating virtual env {self.target}...")
+            self.create_virtual_env()
+            assert self.context is not None
```

Comment:
since `context` field is for cache, it makes sense to create accessor method like `get_context` that will check run `create_virtual_env` if it is empty and always return a context.

Reply:
yes, makes sense, this means we won't need to call `create_virtual_env` directly now for creating the env since `get_context` will handle that for us.

Reply:
actually, I have removed the usage of `context` altogether, so now, we don't have to deal with it + there is no issue of `protected access` as well.